### PR TITLE
Armbench metric dict

### PR DESCRIPTION
**Pull request by RanHomri**
modify metrics and metadata calculation to dictionary

### CODE REVIEW

File: examples/armbench-segmentation/leap.yaml
Comment (Chen-Roth) [inline examples/armbench-segmentation/leap.yaml:1]: add also object_detection.yml

**Review by Chen-Roth — COMMENTED** [review]
Change also in the readme file from cli-go to leap-cli in all places (: 
LSTM

File: examples/armbench-segmentation/armbench_segmentation/metrics.py [hunk 1]
```diff
@@ -99,16 +107,19 @@ def metric_small_bb_in_under_segment(image: tf.Tensor, y_pred_bb: tf.Tensor, y_p
...
-        matches_count = th_arr.astype(int).sum(axis=-1)
-        relevant_bbs = np.argwhere(matches_count > 1)[..., 0]  # [Indices of bbs]
-        relevant_gts = np.where(np.any((th_arr)[relevant_bbs], axis=0))[0]  # [Indices of gts]
-        bb_gt_object, _ = get_mask_list(bb_gt, None, is_gt=True)
-        new_gt_objects = remove_label_from_bbs(bb_gt_object, "Tote", "gt")
-        new_bb_array = [new_gt_objects[i] for i in relevant_gts]
-        for j in range(len(new_bb_array)):
-            if new_bb_array[j].width * new_bb_array[j].height < CONFIG["SMALL_BBS_TH"]:
-                has_small_bbs[i] = 1.
+        if len(ioas) > 0:
```
Comment (yotamazriel) [inline examples/armbench-segmentation/armbench_segmentation/metrics.py:110]: What is `ioas`?

File: examples/armbench-segmentation/leap_binder.py [hunk 2]
```diff
@@ -336,6 +336,73 @@ def count_small_bbs(idx: int, data: PreprocessResponse) -> float:
...
+        "duplicate_bb": count_duplicate_bbs,
+        "small_bbs_number": count_small_bbs,
+        "count_total_obj_bbox_occlusions": get_obj_bbox_occlusions_count,
+        "avg_obj_bbox_occlusions": get_obj_bbox_occlusions_avg,
+        "count_obj_mask_occlusions": get_obj_mask_occlusions_count
+    }
+    res = dict()
+    for func_name, func in metadata_functions.items():
+        res[func_name] = func(idx, data)
+    return res
```
Comment (yotamazriel) [inline examples/armbench-segmentation/leap_binder.py:370]: Why are we doing it?

File: examples/armbench-segmentation/leap_binder.py [hunk 3]
```diff
@@ -336,6 +336,73 @@ def count_small_bbs(idx: int, data: PreprocessResponse) -> float:
...
+    metric_functions = {
+        "Regression_metric": regression_metric,
+        "Classification_metric": classification_metric,
+        "Objectness_metric": object_metric,
+        "Mask_metric": mask_metric,
+    }
+    res = dict()
+    for func_name, func in metric_functions.items():
+        res[func_name] = func(bb_gt, detection_pred, mask_gt, segmentation_pred)
+    return res
```
Comment (yotamazriel) [inline examples/armbench-segmentation/leap_binder.py:384]: This way, we are calculating many calculations four times; let's get into those functions and make sure we run efficiently.

File: examples/armbench-segmentation/leap_binder.py [hunk 4]
```diff
@@ -336,6 +336,73 @@ def count_small_bbs(idx: int, data: PreprocessResponse) -> float:
...
+        "Under_Segmented_Instances_count": under_segmented_instances_count,
+        "Average_segments_num_Over_Segmented": average_segments_num_over_segment,
+        "Average_segments_num_Under_Segmented": average_segments_num_under_segmented,
+        "Over_Segment_confidences": over_segment_avg_confidence
+    }
+    res = dict()
+    for func_name, func in metric_functions.items():
+
+        res[func_name] = func(image, y_pred_bb, y_pred_mask, bb_gt, mask_gt)
+    return res
```
Comment (yotamazriel) [inline examples/armbench-segmentation/leap_binder.py:403]: Same here

File: examples/armbench-segmentation/leap_binder.py [hunk 5]
```diff
@@ -363,42 +432,8 @@ def count_small_bbs(idx: int, data: PreprocessResponse) -> float:
...
-leap_binder.set_metadata(get_object_avg_instance_percent, DatasetMetadataType.float, "object_avg_instance_size")
-leap_binder.set_metadata(get_object_std_instance_percent, DatasetMetadataType.float, "object_std_instance_size")
-leap_binder.set_metadata(bbox_num, DatasetMetadataType.float, "bbox_number")
-leap_binder.set_metadata(get_avg_bb_area, DatasetMetadataType.float, "bbox_area")
-leap_binder.set_metadata(get_avg_bb_aspect_ratio, DatasetMetadataType.float, "bbox_aspect_ratio")
-leap_binder.set_metadata(get_background_percent, DatasetMetadataType.float, "background_percent")
-leap_binder.set_metadata(count_duplicate_bbs, DatasetMetadataType.int, "duplicate_bb")
-leap_binder.set_metadata(count_small_bbs, DatasetMetadataType.int, "small bbs number")
-leap_binder.set_metadata(get_obj_bbox_occlusions_count, DatasetMetadataType.float, "count_total_obj_bbox_occlusions")
-leap_binder.set_metadata(get_obj_bbox_occlusions_avg, DatasetMetadataType.int, "avg_obj_bbox_occlusions")
```
Comment (yotamazriel) [inline examples/armbench-segmentation/leap_binder.py:403]: Nice!

**Review by yotamazriel — CHANGES_REQUESTED** [review]

**Pull request closed without merging**